### DISCUSSION TITLE
Why TWO "about.py" ?

### DISCUSSION THREAD

**Discussion by anneranch** [NanoVNA-Saver/nanovna-saver#650] [Q&A]
EDIT 
Can somebody PLEASE  explain this mess with variety of "versions / releases / URL etc. " references ? 

also 
could  next "release" of nanoVNASaver  add some explanations to confusing variety of "versions"?
I cannot get current "version" to get pass "_version"  ( in one of the about.py code ) -  it does not exist. 
PyCharm tells me that it cannot be completed - depending on how it is retrieved from the cloned file. 
Since that is not my primary goal - verifying "versions"  -   I am busy learning how to  bypass code I do not need as of now. 

Addendum 
Since  this is supposedly "discussion" ...
After "digging into this"  - it definitely needs somebody familiar with all those versions , URL  etc references. 

These - for example - are "direct" (hard coded )  references to repositories , of different   purposes, then when I use Eclipse I  / it adds stuff   using  local , cloned  SAME repositories... Maybe that is why I get "circular referenced ". 
I suspect  Eclipse is actually reading the "git"  - since after I modified  my local "about.py" source ) I added DEBUG word, then  I do not see it.  

VERSION_URL = (
    "https://raw.githubusercontent.com/"
    "NanoVNA-Saver/nanovna-saver/master/NanoVNASaver/About.py"
)
version  = 1
INFO_URL = "https://github.com/NanoVNA-Saver/nanovna-saver"
INFO = f"""NanoVNASaver {version}

DEBUG Copyright (C) 2019, 2020 Rune B. Broberg
Copyright (C) 2020ff NanoVNA-Saver Authors

This program comes with ABSOLUTELY NO WARRANTY
This program is licensed under the GNU General Public License version 3

See {INFO_URL} for further details.
"""

RELEASE_URL = "https://github.com/NanoVNA-Saver/nanovna-saver"

**Comment by t52ta6ek** [comment]
Greetings @anneranch -  I could not figure this out either but after just submitting a pull request to submit a fix for checking for updates, I suspect these two About.py files had something to do with a legacy way of determing what the latest version was. The application version number must have been originally stored in the past in that VERSION_URL About.py.  The other Windows/About.py was used handle the actual work of creating the About-window. And Version.py contained some helper functionality to massage the version string into its components.  Maybe at that time it was thought that other source components beyond Windows/About.py might also need to import those URL references....  I briefly contemplated consolidating About.py and Windows/About.py and Version.py but thought maybe this is an action best left to the maintainers to decide. BR!! Tom

**Comment by anneranch** [comment]
I  have quit worrying about "about(s)" - since it really does not matter to
me.
My initial goal was and still is to improve the tdr option. It works OK ,
but it needs better "configuration / verification" of the scan frequency in
use.
After struggling with Python and current nanoVNASaver code,  it is my
opinion it has too many cooks and not enough
description what to code does. In my opinion , it should have never been
coded in Python. Especially when it is so GUI centered. Python "by
definition"  and the majority of usage is "command  language". It is suited
for "one line code " such as "Hello word". nanoVNASaver is "collector of
data '' coming from nanoVNA and as I see it, it  does very minimal
processing of such data, the majority of processing is GUI oriented.
Hope nobody gets offended by my opinions, and let's just leave it be .

**Reply by zarath** [reply]
Yes, and because of that definition of Python, Dropbox was completely written as one-Liners and Hello World-Applications.
Also Cura is only a small hobby GUI-App.

Seriously, even if the GUI part in Nanovna-Saver isn't the nicest (mostly due to much cut and paste programming during project start), python is a first class choice for Qt-Applications. In my opinion it fits even better to the Qt-Gui concepts as c++ (which requires moc for it's signal/slot implementation).

Further the developer base for python is bigger for python as for c++ nowerdays. Another advantage of python is that it is available on almost every platform. Which GUI-Framework would have been your choice to develop something that runs on a Raspberry Pi, a MacBook, Linux-Devices and Windows?

And yes, TDR could be improved by more automatic reconfiguration of sweep parameters, but even this is hard, as one don't even know the hardware specs of the connected VNAs. For example see: https://github.com/NanoVNA-Saver/nanovna-saver/issues/534#issuecomment-1662670103
 
And regarding versions and updates, I personally never would rely on in program checks, as I didn't expect URLs to be stable. So any change in the software origin will break automated checks anyhow.

**Comment by anneranch** [comment]
Holger Müller Since there is no intention or real need to change
nanoVNASaver outlook , let's analyze the "Python was a better / good
choice" opinion. Especially when you insert the "Qt " in a sentence.  When
I decided that my (hobby) applications would benefit from a better GUI tool
-  I switched to Qt. Then I got into a not so friendly "argument " that it
is not IDE,  but a library.
( After trying to build Bluetooth CAT app  I found it is a lousy Bluetooth
/ bluez based library , but that is a different story)
What we have is "there is relatively unknown data source  - nanoVNA - and
nanoVNASaver (primary) purpose is to display such data". That to me reads -
use GUI tool to accomplish that, but since it was easier to "cut and paste
"   - more than a dozen of  "code copiers / writers" used Python...( not
100% true since PyQt5/6 snuck in ...)  Almost end of the story.
It was my choice to "reverse engineer " nanoVNASaver and being totally
ignorant of Python - the process turned out to be hours of frustration,
mainly because the "code"  is 98.5 %  (sarcasm) void of comments. But it
was my choice and I have learned that any detailed commentaries would only
create bad feelings , at minimum.

It is my conclusion / opinion -
Attempt to "process" nanoVNASaver "mixed tools  / undocumented " Python/
PyQtx code  in QtCreator ( not IDE !)
and / or in PyCharm - IDE -  is asking for punishment ...
